File: python-protobuf/write_message.py
#! /usr/bin/python

from build import addressbook_pb2
import sys

# Fill in a Person message based on user input
def PromptForAddress(person):
    person.id = int(input("Enter person ID number: "))
    person.name = input("Enter name: ")

    email = input("Enter email address (blank for none): ")
    if email != "":
        person.email = email

    while True:
        number = input("Enter a phone number (or leave blank to finish): ")
        if number == "":
            break

        phone_number = person.phones.add()
        phone_number.number = number

        type = input("Is this a mobile, home or work phone? ")
        if type == "mobile":
            phone_number.type = addressbook_pb2.Person.PhoneType.MOBILE
        elif type == "home":
            phone_number.type = addressbook_pb2.Person.PhoneType.HOME
        elif type == "work":
            phone_number.type = addressbook_pb2.Person.PhoneType.WORK
        else:
            print("Unknown phone type; leaving as default value")

# Main procedure: 
#    Read the entire address book from a file
#    Add one person based on user input, then write it back out to the same file

if len(sys.argv) != 2:
    print("Usage:", sys.argv[0], "ADDRESS_BOOK_FILE")
    sys.exit(-1)

# Create an address book (protobuf message)
address_book = addressbook_pb2.AddressBook()

# Read the existing address book
try:
    with open(sys.argv[1], "rb") as f:
        address_book.ParseFromString(f.read())
except IOError:
    print(sys.argv[1] + ": File not found. Creating a new file.")

# Add an address
PromptForAddress(address_book.people.add())

#Write the new address book back to disk
with open(sys.argv[1], "wb") as f:
    f.write(address_book.SerializeToString())

File: csharp-protobuf/WriteMessage.cs
using System;
using System.IO;

using Google.Protobuf;
using static Proto.Addressbook.Person.Types;

namespace Proto.Addressbook
{
    class WriteMessage
    {
        private static Person PromptForAddress(TextReader input, TextWriter output)
        {
            Person person = new Person();

            output.Write("Enter person ID: ");
            person.Id = int.Parse(input.ReadLine());

            output.Write("Enter name: ");
            person.Name = input.ReadLine();

            output.Write("Enter email address (blank for none): ");
            string email = input.ReadLine();
            if (email.Length > 0)
            {
                person.Email = email;
            }

            while (true)
            {
                output.Write("Enter a phone number (or leave blank to finish): ");
                string number = input.ReadLine();
                if (number.Length == 0)
                {
                    break;
                }

                PhoneNumber phoneNumber = new PhoneNumber { Number = number };

                output.Write("Is this a mobile, home, or work phone? ");
                String type = input.ReadLine();
                switch(type)
                {
                    case "mobile":
                        phoneNumber.Type = PhoneType.Mobile;
                        break;
                    case "home":
                        phoneNumber.Type = PhoneType.Home;
                        break;
                    case "work":
                        phoneNumber.Type = PhoneType.Work;
                        break;
                    default:
                        output.Write("Unknown phone type. Using default.");
                        break;
                }
                person.Phones.Add(phoneNumber);
            }
            return person;
        }

        public static int Main(string[] args)
        {
            if (args.Length != 1)
            {
                Console.Error.WriteLine("Usage: WriteMessage ADDRESS_BOOK_FILE");
                return -1;
            }

            AddressBook addressBook;

            if (File.Exists(args[0]))
            {
                using (Stream file = File.OpenRead(args[0]))
                {
                    addressBook = AddressBook.Parser.ParseFrom(file);
                }
            }
            else
            {
                Console.WriteLine("{0}: File not found. Creating a new file.", args[0]);
                addressBook = new AddressBook();
            }

            // Add an address
            addressBook.People.Add(PromptForAddress(Console.In, Console.Out));

            // Write the new address book back to disk
            using (Stream output = File.OpenWrite(args[0]))
            {
                addressBook.WriteTo(output);
            }
            return 0;
        }
    }
}
